lib/THC/THCTensorInfo.cuh

```diff
@@ -223,6 +233,23 @@ TensorInfo<T, IndexType>::collapseDims(int excludeDim) {
   return returnDim;
 }
 
+template <typename T, typename IndexType>
+TensorInfo<T, IndexType>
+TensorInfo<T, IndexType>::newNarrow(int dimension, IndexType firstIndex, IndexType size) const {
+  IndexType sz[MAX_CUTORCH_DIMS];
```

Comment:
This would now be a device function as well, and MAX_CUTORCH_DIMS is so large now too, I think that this would either be assigned to many registers or to local memory, negatively impacting perf.

Probably what needs to happen is add `int MaxDims = MAX_CUTORCH_DIMS` as an additional template argument to TensorInfo, and then specialize the kernel to restrict # of dims to 2 - 4 or something for the parallel version.

The excessive size of MAX_CUTORCH_DIMS would help your # of kernel launch problems as well.

Maybe it's only worthwhile to call the all-in-one kernel in the case that all of the tensors being cat'ed are contiguous and are dim-4 or less, so you don't need a stride array for any of them?